# **Problem Statement**: Create a simple recommendation system that suggests items based on their preferences. You can use techniques like collaborative filtering or content-based filtering to recommend movies, books, or products to users.

In [2]:
# Import necessary libraries and install surprise
#!pip install scikit-surprise

In [3]:
import pandas as pd
from surprise import Dataset, Reader
from surprise.model_selection import train_test_split
from surprise import KNNBasic
from surprise import accuracy

# **Load and preprocess the data**

In [20]:
# In this Project, we'll use the MovieLens dataset.
file_path = "/content/MovieLens.csv"


In [21]:
# Load the dataset using pandas
data_df = pd.read_csv(file_path)

In [27]:
data_df.head()

userId  movieId  rating
0       1      296     5.0
1       1      306     3.5
2       1      307     5.0
3       1      665     5.0
4       1      899     3.5

In [28]:
# Define the Reader object
reader = Reader(rating_scale=(1, 5))

In [29]:
# Load the data from the Pandas DataFrame into a Dataset object
data = Dataset.load_from_df(data_df[['userId', 'movieId', 'rating']], reader)

In [30]:
# Split the data into training and testing sets
trainset, testset = train_test_split(data, test_size=0.2, random_state=42)


# **Build and train the recommendation model**
we can choose different algorithms for collaborative filtering, like SVD, KNNBasic, etc.

In [31]:
model = KNNBasic(sim_options={'user_based': True})  # User-based collaborative filtering

In [32]:
# Fit the model on the training data
model.fit(trainset)

Computing the msd similarity matrix...
Done computing similarity matrix.


# **Make recommendations for a user**

In [43]:
user_id = input("Enter The User!")
print("you are looking for top 10 recommendations for",user_id)
n_recommendations = 10  # Number of recommendations to generate

Enter The User!5
you are looking for top 10 recommendations for 5


In [44]:
# Get a list of all item IDs
all_item_ids = data.build_full_trainset().all_items()

In [45]:
# Remove items that the user has already rated
items_to_predict = [item_id for item_id in all_item_ids if item_id not in trainset.ur[user_id]]

In [46]:
# Make predictions for the user
predictions = [model.predict(user_id, item_id) for item_id in items_to_predict]

In [47]:
# Sort the predictions by estimated rating in descending order
predictions.sort(key=lambda x: x.est, reverse=True)


In [48]:
# Get the top N recommendations
top_n_recommendations = predictions[:n_recommendations]

In [49]:
# Print the top recommendations
print(f"Top {n_recommendations} Recommendations for User {user_id}:")
for i, prediction in enumerate(top_n_recommendations):
    print(f"{i + 1}: {prediction.iid} (Estimated Rating: {prediction.est:.2f})")

Top 10 Recommendations for User 5:
1: 0 (Estimated Rating: 3.54)
2: 1 (Estimated Rating: 3.54)
3: 2 (Estimated Rating: 3.54)
4: 3 (Estimated Rating: 3.54)
5: 4 (Estimated Rating: 3.54)
6: 5 (Estimated Rating: 3.54)
7: 6 (Estimated Rating: 3.54)
8: 7 (Estimated Rating: 3.54)
9: 8 (Estimated Rating: 3.54)
10: 9 (Estimated Rating: 3.54)


# **Evaluate the model**

In [41]:
# You can evaluate the model's performance on the test set

test_predictions = model.test(testset)
rmse = accuracy.rmse(test_predictions)
mae = accuracy.mae(test_predictions)
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")

RMSE: 0.9000
MAE:  0.6829
Root Mean Squared Error (RMSE): 0.90
Mean Absolute Error (MAE): 0.68


# *****In summary, an RMSE of 0.9 and an MAE of 0.68 are reasonable starting points, but whether My Model is considered "good" depends on the factors Like user satisfaction and baselinecomparison. It's also common to continuously refine and improve recommendation models to achieve better performance if needed.*****